---
title: Multiomics reveals persistence of obesity-associated adipose tissue immune
  cell phenotypes during weight loss and subsequent weight regain
author: "Matthew A. Cottam"
subtitle: "Figure 2"
output:
  html_document:
    toc: yes
    df_print: paged
  html_notebook:
    theme: journal
    fig_caption: yes
    toc: yes
---

# Setup

```{r prep, warning=FALSE, message=FALSE}
library(Seurat)
library(ape)
library(RColorBrewer)
library(ggplot2)

#Data file
data.integrated <- readRDS("../Inputs/IntegratedData.rds")

#Color panels
maccols <- brewer.pal(n=8, name="Blues")[c(-1,-3,-5,-7)]
monocols <- c("#ff8ade","#e324ad")
dccols <- brewer.pal(n=9, name="Greens")[-1]
tcols <- brewer.pal(n=8, name="Reds")[-1]
nkcols <- c("#876149","#6e3f22")
bcols <- brewer.pal(n=4, name="Purples")[-1]
othcols <- c("#71a157","#00c5cc","#a18e25","#b282b3")
strcols <- brewer.pal(n=4, name="Oranges")[-1]
wccols = c("#878787", "#518db6","#94cc73","#e96b53")

cols <- c(maccols,monocols,dccols,tcols,nkcols,bcols,othcols,strcols)

llcols <- c("#4292C6","#ff8ade","#238B45","#EF3B2C","#876149","#9E9AC8","#71a157","#00c5cc","#a18e25","#b282b3","#FD8D3C")
```

# Figure 2

Figure 2 introduces our single cell sequencing approach and the simple classification of cell types using cell-type exclusive markers and targeted surface protein analysis.

## Figure 2A

Schematic of CITE-Seq approach

## Figure 2B

```{r Fig2Bll, fig.width=10, fig.height=10, echo=TRUE, warning=FALSE, message=FALSE}
DimPlot(data.integrated, group.by="lowlevel2", cols=llcols) + NoLegend() + labs(title="")
```

```{r Fig2B, fig.width=10, fig.height=10, warning=FALSE, message=FALSE}
DimPlot(data.integrated, group.by="highlevel2", cols=cols) + NoLegend() + labs(title="")
#For labels
DimPlot(data.integrated, group.by="highlevel2", cols=cols) + labs(title="") 
```

## Figure 2C

```{r Fig2C, caption="Cell type markers from gene expression data", fig.width=6, fig.height=6, echo=TRUE, warning=FALSE, message=FALSE}
Idents(data.integrated) <- data.integrated$lowlevel2
DefaultAssay(data.integrated) <- "RNA"
VlnPlot(data.integrated, 
        features=c("Fcgr1","Ear2","Clec9a","Cd3d","Klrb1c","Ms4a1","Jchain","Mcpt4","S100a8","Gata3","Col3a1"), 
        stack=T, 
        flip=T,
        fill.by="ident",
        cols = llcols) +
  labs(x="") + 
  NoLegend()
```

## Figure 2D

```{r Fig2D, caption="Cell type markers from CITE-seq", fig.width=6, fig.height=6, warning=FALSE, message=FALSE}
DefaultAssay(data.integrated) <- "ADT"
Idents(data.integrated) <- data.integrated$lowlevel2
VlnPlot(data.integrated, 
        features=c("MAC2","CD64","CD11b","CD11c","CD80","CD3","CD4","CD19","CD39","NK1.1"), 
        stack=T, 
        flip=T,
        fill.by="ident",
        cols = llcols) +
  labs(x="") + 
  NoLegend()
```

## Figure 2E
```{r Fig2E, fig.width=9, fig.height=6, warning=FALSE, message=FALSE}
DefaultAssay(data.integrated) <- "integrated"
Idents(data.integrated) <- data.integrated$highlevel2
data.integrated <- BuildClusterTree(object=data.integrated,
                 assay="integrated")
data.tree <- Tool(object = data.integrated, slot = "BuildClusterTree")
ape::plot.phylo(x = data.tree, direction = "downwards", label.offset=60, cex=0.8)
```

# Session Info

```{r si}
sessionInfo()
```

